main/lsp/requests/completion.cc

```diff
@@ -7,50 +7,122 @@ using namespace std;
-                                  make_move_iterator(other.second.end()));
+namespace {
+
+core::NameRef originalName(const core::GlobalState &gs, core::NameRef name) {
+    switch (name.data(gs)->kind) {
+        case core::NameKind::UTF8:
+            return name;
+        case core::NameKind::UNIQUE:
+            return originalName(gs, name.data(gs)->unique.original);
+        case core::NameKind::CONSTANT:
+            return originalName(gs, name.data(gs)->cnst.original);
     }
-    return std::move(first);
+}
+
+vector<core::SymbolRef> ancestorsImpl(const core::GlobalState &gs, core::SymbolRef sym, vector<core::SymbolRef> &&acc) {
+    // The implementation here is similar to Symbols::derivesFrom.
+    ENFORCE(sym.data(gs)->isClassOrModuleLinearizationComputed());
+    acc.emplace_back(sym);
+
+    for (auto mixin : sym.data(gs)->mixins()) {
+        acc.emplace_back(mixin);
+    }
+
+    if (sym.data(gs)->superClass().exists()) {
+        return ancestorsImpl(gs, sym.data(gs)->superClass(), move(acc));
+    } else {
+        return move(acc);
+    }
+}
+
+// Basically the same as Module#ancestors from Ruby--but don't depend on it being exactly equal.
+// For us, it's just something that's vaguely ordered from "most specific" to "least specific" ancestor.
+vector<core::SymbolRef> ancestors(const core::GlobalState &gs, core::SymbolRef receiver) {
+    return ancestorsImpl(gs, receiver, vector<core::SymbolRef>{});
+}
+
+struct SimilarMethod final {
+    int depth;
+    core::SymbolRef receiver;
+    core::SymbolRef method;
+
+    // Populated later
+    core::TypePtr receiverType = nullptr;
+    shared_ptr<core::TypeConstraint> constr = nullptr;
 };
 
-UnorderedMap<core::NameRef, vector<core::SymbolRef>> findSimilarMethodsIn(const core::GlobalState &gs,
-                                                                          core::TypePtr receiver, string_view name) {
-    UnorderedMap<core::NameRef, vector<core::SymbolRef>> result;
-    typecase(
-        receiver.get(),
-        [&](core::ClassType *c) {
-            const auto &owner = c->symbol.data(gs);
-            for (auto member : owner->membersStableOrderSlow(gs)) {
-                auto sym = member.second;
-                if (sym.data(gs)->isMethod() && hasSimilarName(gs, sym.data(gs)->name, name)) {
-                    result[sym.data(gs)->name].emplace_back(sym);
-                }
-            }
-            for (auto mixin : owner->mixins()) {
-                result = mergeMaps(std::move(result),
-                                   findSimilarMethodsIn(gs, core::make_type<core::ClassType>(mixin), name));
+// First of pair is "found at this depth in the ancestor hierarchy"
+// Second of pair is method symbol found at that depth, with name similar to prefix.
+vector<SimilarMethod> similarMethodsForClass(const core::GlobalState &gs, core::SymbolRef receiver,
+                                             string_view prefix) {
+    auto result = vector<SimilarMethod>{};
```

Comment:
You can just write `vector<SimilarMethod> result;` and it'll invoke the default constructor.

Reply:
I like using `auto` because i find it's easier to scan for where a variable is defined if there's not a long typename like `vector<SimilarMethod>` on the left side. I can switch to the other way if you prefer.

Reply:
You can stick with `auto`, but note that we use the convention I suggested in most other locations of the code.